File: Metaflow/FeatureClientGrain.fs
namespace Metaflow

open System
open System.Threading.Tasks
open Metaflow
open Orleans
open FSharp.Control.Tasks
open Orleans.Concurrency
open Serilog

type IFeatureClientGrain =
    inherit IGrainWithStringKey
    abstract Delete<'model> : string * bool -> Task<Result<unit, WorkflowFailure>>

type FeatureClient(clusterClient: IClusterClient) =
    member this.Delete<'model>(aggregateRootId: string, awaitState: bool) =
        let grain =
            clusterClient.GetGrain<IFeatureClientGrain>(Guid.NewGuid().ToString())

        grain.Delete<'model>(aggregateRootId, awaitState)

[<StatelessWorker>]
type FeatureClientGrain(clusterClient: IClusterClient, workflows: Workflow seq) =
    inherit Grain()

    let workflowMap =
        workflows
        |> Seq.map (fun f -> ((f.Feature.Model.FullName, f.Feature.Operation), f))
        |> Map.ofSeq

    interface IFeatureClientGrain with
        member this.Delete<'model>(aggregateRootId: string, awaitState: bool) =
            task {
                let workflowOption =
                    workflowMap
                    |> Map.tryFind (typeof<'model>.FullName, Operation.DELETE)

                return!
                    match workflowOption with
                    | None -> task { return Error(WorkflowFailure.FeatureFailure(FeatureFailure.NotFound)) }
                    | Some workflow ->
                        let call =
                            { Feature = workflow.Feature
                              Input = Id(OwnedValueId { AggregateRootId = aggregateRootId })
                              AwaitState = awaitState }

                        task {
                            return!
                                Workflows.run<unit, Delete, 'model>
                                    workflow
                                    call
                                    clusterClient
                                    { RequestId = Guid.NewGuid().ToString() }
                                    Log.Logger
                        }
            }

File: Metaflow.Orleans.Grains/FeatureClientGrain.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Core;
using Orleans;
using Orleans.Concurrency;

namespace Metaflow.Orleans
{
    [StatelessWorker]
    public class FeatureClientGrain : Grain, IFeatureClientGrain
    {
        private readonly IClusterClient _clusterClient;
        private readonly FSharpMap<Tuple<string, Operation>, Workflow> _workflowMap;

        public FeatureClientGrain(IClusterClient clusterClient, IEnumerable<Workflow> workflows)
        {
            _clusterClient = clusterClient;
            _workflowMap =
                new FSharpMap<Tuple<string, Operation>, Workflow>(
                    workflows.Select(w =>
                        Tuple.Create(
                            Tuple.Create(w.Feature.Model.FullName, w.Feature.Operation), w)));
        }

        public Task<FSharpResult<Unit, WorkflowFailure>> Delete<TModel>(string id, bool awaitState)
        {
            return Features.Delete<TModel>.Execute(_workflowMap, _clusterClient, id, awaitState);
        }
    }
}